### DISCUSSION TITLE
Errors while installing packages using different versions of create-toolpad-app in a monorepo

### DISCUSSION THREAD

**Discussion by iamlvv** [mui/toolpad#3692] [Help and support] [answered]
Hi, I am currently working on a monorepo project. In this project, I already created `workspaceOld` (in folder packages) with `create-toolpad-app` version `0.1.44`. After that, I upgrade `create-toolpad-app` to `0.2.0`(the latest) and create new `workspaceNew`.

The error starts here. When I run the `workspaceNew` in dev, the components do not have the MUI design but HTML instead. For example, In editor, I only see the button in HTML design, instead of MUI Button. However, when I click on Preview, the MUI Button will replace the HTML button.

Expect:
<img width="1402" alt="image" src="https://github.com/mui/mui-toolpad/assets/48018773/af0a55b5-9590-482e-88e2-03b4d1b9df17">
Turn out to be like this:
<img width="1483" alt="image" src="https://github.com/mui/mui-toolpad/assets/48018773/fc52888f-418f-4fbc-97d0-2f131671d2f6">

However, when I change the script in `package.json` file in `workspaceNew`, from:
```
"scripts": {
    "dev": "toolpad-studio dev",
    "build": "toolpad-studio build",
    "start": "toolpad-studio start"
  },
```
to 
```
"scripts": {
    "dev": "toolpad dev",
    "build": "toolpad build",
    "start": "toolpad start"
  },
```
The MUI Design Components show up and work properly, but the URL for editor changes from this:
```
http://localhost:3000/prod/editor/app/pages/page
```
to this:
```
http://localhost:3000/_toolpad/app/pages/page
```
Can you share more on this and how I can work around this problem?

**Comment by Janpot** [comment]
Thanks @iamlvv. The quickest way for us to assess the problem you run into would be through a minimal reproduction. Would you be able to create a small workspace that demonstrates the issue and share it as a git repo, or even just a zip file (you can remove the mode_modules folder to keep the size down)?

From your description it looks like It's unexpectedly sharing some dependencies between the two projects. In general I wouldn't recommend running multiple versions of toolpad in the same workspace.

**Reply by iamlvv** [reply]
Hi, when I create another pet project, I see that it works properly even when I switch from 0.1.44 to 0.2.0.

However, coming back to my monorepo project, I find out that the package.json of the `workspaceOld` is like this:
```
{
  "name": "test-app-1",
  "version": "0.1.0",
  "scripts": {
    "dev": "toolpad dev",
    "build": "toolpad build",
    "start": "toolpad start"
  },
  "dependencies": {
    "@mui/toolpad": "^0.1.45"
  }
}

```
 Meanwhile, the package.json of the `workspaceNew` has dependency like this:
```
{
  "name": "test-app-2",
  "version": "0.1.0",
  "scripts": {
    "dev": "toolpad-studio dev",
    "build": "toolpad-studio build",
    "start": "toolpad-studio start"
  },
  "dependencies": {
    "@toolpad/studio": "latest"
  }
}
```
I see that @mui/toolpad has been deprecated. Is it possible the root cause?

**Reply by Janpot** [reply]
> I see that @mui/toolpad has been deprecated. Is it possible the root cause?

`@mui/toolpad` has been renamed to `@toolpad/studio`. I don't expect there to be breaking changes. Have you tried upgrading your old project to `@toolpad/studio`? (you'll have to update the npm scripts from `toolpad` to `toolpad-studio`)

Toolpad studio has been build on top of `vite`, and we know `vite` has some behavior that affect global setup (e.g. dependency optimization). I could imagine the setup getting a bit confused when multiple versions of Toolpad are mixed in the same workspace. I could take a look if I had access to a reproduction.

**Reply by iamlvv** [reply]
Hi, I have just created a project to reproduce the issue. You can refer to [this link](https://github.com/iamlvv/test-mui-toolpad.git). 

I have also written steps I have done when installing and creating apps with `create-toolpad-app`.

**Reply by Janpot** [reply]
I have a PR that could fix your issues: https://github.com/mui/mui-toolpad/pull/3695

To try it out you could change the `@toolpad/studio` dependency in `workspaceNew` as follows:

```json
"@toolpad/studio": "https://pkg.csb.dev/mui/mui-toolpad/commit/bff29fb7/@toolpad/studio"
```

I want to stress. Multiple different versions of Toolpad in a monorepo is not something we explicitly support. I advise you to instead upgrade your projects that use `@mui/toolpad` to `@toolpad/studio` and use a single version of toolpad across the codebase.

**Reply by iamlvv** [reply]
Hi, thank you for that. I will check it out soon. However, can you share the root cause with me?

**Reply by Janpot** [reply, marked as answer]
Multiple versions of `@emotion/react` in a page and multiple versions of `@tanstack/react-query` in a page make producers and consumers of certain React context not match up with each other. 
For now I forced depuplicating these packages in the toolpad bundler, but I can't guarantee this won't start breaking again some day with a different package,
`npm` dependency hoisting makes modules end up being resolved from places they shouldn't be resolved from. Converting your workspace to a `pnpm` workspace also fixes the issue for me as it creates a much more predictable node_modules structure to resolve from.

**Reply by iamlvv** [reply]
Hi @Janpot, I have checked and it has worked properly. Thank you so much for your help. For the long-term solution, I may upgrade the projects but in practice, sometimes we cannot upgrade immediately. Converting to `pnpm` is fine, I may stick this way.
